examples/ssr-data-router/src/entry.server.tsx

```diff
@@ -9,15 +9,15 @@ import {
 import { routes } from "./App";
 
 export async function render(request: express.Request) {
-  let { query } = createStaticHandler(routes);
+  let { query, dataRoutes } = createStaticHandler(routes);
```

Comment:
I tried my hand at baking the lazy route module mapping code into react-router and this is where I got stuck.

This `createStaticHandler` creates the `query` function which resolves the lazy route modules, which means it needs React-specific knowledge for mapping `default` to `element` and `ErrorBoundary` to `errorElement`. The challenge is that consumers are importing `createStaticHandler` directly from `@remix-run/router` which has no knowledge of React. Does this mean react-router needs its own `createStaticHandler` function which internally would pass a React-specific route module mapping function down to the `createStaticHandler` in `@remix-run/router`?

Reply:
> Does this mean react-router needs its own `createStaticHandler` function which internally would pass a React-specific route module mapping function down to the `createStaticHandler` in `@remix-run/router`?

Yeah that's what I've been thinking too.  

For client-side, I think we can do it inside `createBrowserRouter` and the end-user will never see it:

```js
function createBrowserRouter(routes, opts) {
  return createRouter({
    ...
    routes: enhanceManualRouteObjects(routes),
    routeMapper: (route) => ({  
      action: route.action,
      loader: route.loader,
      element: <route.default />,
      errorElement: <route.ErrorBoundary />,
      hasErrorBoundary: route.ErrorBoundary != null,
    }),
  }).initialize();
}
```

If we go this route we may actually be able to eliminate a tree-walk and leverage `routeMapper` inside `convertRoutesToDataRoutes` and eliminate `enhanceManualRouteObjects` since the only thing that does is assign `hasErrorBoundary`.  This would likely require the `routeMapper` code above to become a tiny bit smarter and more defensive.

For static cases, maybe `react-router` could export this `routeMapper` and we could pass it to `createStaticHandler` as well?  Right now users are required to include `hasErrorBoundary` on their `routes` passed to `createStaticHandler`, but this could eliminate that if we leverage `routeMapper` inside `convertRoutesToDataRoutes`.  

```js
import { routeMapper } from `react-router`;

let staticHandler = createStaticHandler(routes, { routeMapper });
```

So `routeMapper` would be used for 2 things:

* On router creation inside `convertRoutesToDataRoutes` to handle `hasErrorBoundary` assignment based on `errorElement`
* Then again for any lazily loaded route

Reply:
We already spoke about this offline, but for anyone following along — this got further simplified since we're not mapping `default`/`ErrorBoundary` exports anymore, so all that's left is calculating the `hasErrorBoundary` flag. `createStaticHandler` and other low-level router APIs now accept a `hasErrorBoundary` _function_ which is passed a route object and returns a boolean. This allows us to write a React-specific adapter while keeping the core agnostic.

Reply:
@brophdawg11 @markdalgleish I see you guys didn't go finally with the `routeMapper` but I wanted to mentioned a use case that needs the mapping.

At [`generouted`](https://github.com/oedotme/generouted), I'm using route-based exports for `element`, `errorElement`, `loader`, etc. and we had to use capitalized export names as a [workaround with Vite HMR to make it functional in dev mode](https://github.com/oedotme/generouted/issues/5), so for example we use `Loader` instead of `loader`

The new route `lazy` API looks great and if I understand correctly it does or will support pre-loading. That would make things much easier overall, but with the current lazy API it wouldn't be possible to use it with `generouted`.

I was wondering if it's possible to have an optional simple route's lazy properties mapping as an option to `createBrowserRouter` for example:

```tsx
type LazyMapping = Partial<Record<'element' | 'loader', string>>

const router = createBrowserRouter(routes, {
  lazyMapping: {
    element: 'default',
    loader: 'Loader',
    // ...
  } 
})
```

Later when accessing the module exports, it can be through `module[lazyMapping.element]` instead of `module.element`.

I'm not sure if this would work or if you guys haven't tried it already, but please let me know what you think, and thanks for the great effort putting all this together! :heart:

Reply:
You can do any mapping you want inside `lazy`:

```js
let routes = [{
  path: '/',
  async lazy() {
    let m = await import('./route');
    return {
      element: m.default,
      loader: m.Loader,
    };
  }
}];
```

And therefore you can write your own mappers:

```js
function mapper(loadModule) {
    let m = await loadModule();
    return {
      element: m.default,
      loader: m.Loader,
    };
}

let routes = [{
  path: '/a',
  lazy: () => mapper(() => import('./a')),
}, {
  path: '/b',
  lazy: () => mapper(() => import('./b')),
}];
```

Reply:
Yeah makes sense, thanks for your reply!